File: 剑指offer/55.2.平衡二叉树.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def isBalanced(self, root: TreeNode) -> bool:
        ans = True

        def help(root):
            nonlocal ans
            if not root:
                return 0

            left_height = help(root.left)
            right_height = help(root.right)

            if abs(left_height - right_height) > 1:
                ans = False

            return 1 + max(left_height, right_height)

        help(root)
        return ans

File: 剑指offer/55.2.平衡二叉树.cpp
#include <iostream>
using namespace std;

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x)
        : val(x)
        , left(NULL)
        , right(NULL)
    {
    }
};

class Solution {
public:
    bool ans = true;

    int help(TreeNode* root)
    {
        if (root == NULL || !ans) {
            return 0;
        }

        int left = help(root->right);
        int right = help(root->left);

        if (abs(left - right) > 1) {
            ans = false;
        }

        return max(left, right) + 1;
    }

    bool isBalanced(TreeNode* root)
    {
        help(root);
        return ans;
    }
};